src/main/scala/is/hail/methods/PCRelate.scala

```diff
@@ -0,0 +1,256 @@
+        val it = gs.hardCallIterator
+        while (it.hasNext) {
+          val gt = it.next()
+          if (gt == -1) {
+            missingIndices += i
+          } else {
+            sum += gt
+            a(i) = gt
+            nNonMissing += 1
+          }
+          i += 1
+        }
+
+        val mean = sum.toDouble / nNonMissing
+
+        for (i <- missingIndices.result()) {
+          a(i) = mean
+        }
+
+        // FIXME: this should probably be a sparse vector
+        new IndexedRow(variantIdxBc.value(v), new DenseVector(a))
+      }
+    }
+    new IndexedRowMatrix(rdd, variants.length, nSamples)
+  }
+
+  /**
+    *  g: SNP x Sample
+    *  pcs: Sample x D
+    *
+    *  result: (SNP x (D+1))
+    */
+  def fitBeta(g: IndexedRowMatrix, pcs: DenseMatrix, blockSize: Int): M = {
+    val aa = g.rows.sparkContext.broadcast(pcs.rowIter.map(_.toArray).toArray)
+    val rdd = g.rows.map { case IndexedRow(i, v) =>
+      val ols = new OLSMultipleLinearRegression()
+      ols.newSampleData(v.toArray, aa.value)
+      val a = ols.estimateRegressionParameters()
+      IndexedRow(i, new DenseVector(a))
+    }
+    dm.from(new IndexedRowMatrix(rdd, g.numRows(), pcs.numCols + 1), blockSize, blockSize)
+  }
+
+  def k1(k2: M, k0: M): M = {
+    1.0 - (k2 :+ k0)
+  }
+
+  def ibs0(vds: VariantDataset, blockSize: Int): M =
+    dm.from(IBD.ibs(vds)._1, blockSize, blockSize)
+
+}
+
+class PCRelate(maf: Double, blockSize: Int) extends Serializable {
+  import PCRelate._
+  import dm.ops._
+
+  require(maf >= 0.0)
+  require(maf <= 1.0)
+  val antimaf = (1.0 - maf)
+  def badmu(mu: Double): Boolean =
+    mu <= maf || mu >= antimaf || mu <= 0.0 || mu >= 1.0
+  def badgt(gt: Double): Boolean =
+    gt != 0.0 && gt != 1.0 && gt != 2.0
+
+  def apply(vds: VariantDataset, pcs: DenseMatrix): Result[M] = {
+    val g = vdsToMeanImputedMatrix(vds)
+    val beta = fitBeta(g, pcs, blockSize)
+
+    val pcsWithIntercept = prependConstantColumn(1.0, pcs)
+
+    val blockedG = dm.from(g, blockSize, blockSize)
+    val mu = ((beta * pcsWithIntercept.transpose) / 2.0)
+
+    val variance = dm.map2 {
+      case (g, mu) =>
+        if (badgt(g) || badmu(mu))
+          0.0
+        else
+          mu * (1.0 - mu)
+    }(blockedG, mu)
+
+    val phi = this.phi(mu, variance, blockedG)
+
+    val k2 = this.k2(phi, mu, variance, blockedG)
+    val k0 = this.k0(phi, mu, k2, blockedG, ibs0(vds, blockSize))
+    val k1 = 1.0 - (k2 :+ k0)
+
+    Result(phi, k0, k1, k2)
+  }
+
+  private def phi(mu: M, variance: M, g: M): M = {
+    val centeredG = dm.map2 { (g,mu) =>
+      if (badgt(g) || badmu(mu))
+        0.0
+      else
+        g - mu * 2.0
+    }(g, mu)
+    val stddev = dm.map(math.sqrt _)(variance)
+
+    ((centeredG.t * centeredG) :/ (stddev.t * stddev)) / 4.0
+  }
+
+  private def k2(phi: M, mu: M, variance: M, g: M): M = {
+    val twoPhi_ii = dm.diagonal(phi).map(2.0 * _)
+    val normalizedGD = dm.map2WithIndex({
+      case (_, i, g, mu) =>
```

Comment:
prefer these on the same line as the opening brace. You've also got an unnecessary set of parens on this map